Extend autopkgtest hints

cpaelzer opened the pull request:
I did intentionally not go deeper as that would then be for the man-ages or the linked wiki entries.
But to list them as options is IMHO helpful to new team members.

bryceharrington commented on PackageTests.md:
```diff
@@ -78,6 +78,28 @@ The command only differs after the `--` part. For example:
+| 4d24cfbe-b6a5-4d84-8c50-b9f025d0dd43 | ubuntu/ubuntu-hirsute-daily-arm64-server-20201124-disk1.img    | active |
+| 1cfeacff-f04a-4bce-ab92-9d8fec7e5edb | ubuntu/ubuntu-hirsute-daily-arm64-server-20201125-disk1.img    | active |
+
+[Finally to run the test on Canonistack](https://wiki.ubuntu.com/ProposedMigration#Reproducing_tests_in_the_cloud) is quite similar to the other invocations. Just replace the autopkgtest-runner with `nova` and adding the `setup-testbed` setup command.
```
The Canonistack web page docs don't mention `nova` or `setup-testbed`.  So might be helpful here to briefly state what the former means and what the latter does.

cpaelzer replied:
Indeed, added

bryceharrington reviewed (changes requested):
I like it, short and sweet.

I do suggest a few changes where a small bit more words may improve the reading from a new user POV.

Thanks for writing this up, you've at least taught me how to run autopkgtests in canonistack. :-D

cpaelzer commented on PackageTests.md:
```diff
@@ -78,6 +78,28 @@ The command only differs after the `--` part. For example:
+This is by far the closest in terms of "similarity" to the real autopkgtests since they also run in such an environment.
+But it needs some preparation. First of all you must have been *unlocked for* and have set up [Canonistack](https://wiki.canonical.com/InformationInfrastructure/IS/CanoniStack-BOS01) for yourself.
+
+As usual with Canonistack you'd need to source your openstack RC file that sets region, auth and other environment variables.
```
> In case someone is new to Canonistack, 'As usual' might sound like it's glossing over a scary amount of technical unknowns. 

Which is exactly what I did :-)
 
> In going through the set up process for Canonistack, you'll have created an openstack RC file that sets region, auth and other environment variables. Go ahead and source this file, if you haven't already.

I still like your words more, adopted - thanks!

cpaelzer commented on PackageTests.md:
```diff
@@ -78,6 +78,28 @@ The command only differs after the `--` part. For example:
+As usual with Canonistack you'd need to source your openstack RC file that sets region, auth and other environment variables.
+Then you'd look for the image you want to boot like:
+
+$ source novarc_bos01
```
Agreed, it helps to keep those in match

cpaelzer commented:
The fixups for @bryce made me made me also improve a few other things here and there.
Thanks for the feedback, I've pushed my cleanups to this branch.
Is there anything else that would need to change or can we merge it this way now?

utkarsh2102 commented on PackageTests.md:
```diff
@@ -78,6 +78,54 @@ The command only differs after the `--` part. For example:
+This is by far the closest in terms of "similarity" to the real autopkgtests since they also run in such an environment.
+But it needs some preparation. First of all you must have been *unlocked for* and have set up [Canonistack](https://wiki.canonical.com/InformationInfrastructure/IS/CanoniStack-BOS01) for yourself.
+
+In going through the set up process for Canonistack, you'll have created an openstack RC file that sets region, auth and other environment variables. Go ahead and source this file, if you haven't already.
```
Maybe it's worth telling that you'll likely get it from the IS team, you just need to place and use the right one?
It's mentioned in the docs, I guess.

cpaelzer replied:
It is mentioned in the linked Canonistack wiki - and since that touches internal processes and the handbook is public I wanted to leave it as I wrote.

utkarsh2102 commented on PackageTests.md:
```diff
@@ -78,6 +78,54 @@ The command only differs after the `--` part. For example:
+In going through the set up process for Canonistack, you'll have created an openstack RC file that sets region, auth and other environment variables. Go ahead and source this file, if you haven't already.
+Then you'd look for the image you want to boot like:
+
+$ source ~/.canonistack/novarc_bos01
+$ openstack image list | grep -i arm64 | grep hirsute
+| 4d24cfbe-b6a5-4d84-8c50-b9f025d0dd43 | ubuntu/ubuntu-hirsute-daily-arm64-server-20201124-disk1.img    | active |
+| 1cfeacff-f04a-4bce-ab92-9d8fec7e5edb | ubuntu/ubuntu-hirsute-daily-arm64-server-20201125-disk1.img    | active |
+
```
This is bad formatting, unfortunately; see the output here: https://github.com/cpaelzer/ubuntu-maintainers-handbook/blob/extend-autopkgtest/PackageTests.md

These two lines should be placed inside backticks or/and you need to leave two spaces at the end of the first line so that MD renders it correctly.

utkarsh2102 commented on PackageTests.md:
```diff
@@ -78,6 +78,54 @@ The command only differs after the `--` part. For example:
+| 1cfeacff-f04a-4bce-ab92-9d8fec7e5edb | ubuntu/ubuntu-hirsute-daily-arm64-server-20201125-disk1.img    | active |
+
+[Finally to run the test on Canonistack](https://wiki.ubuntu.com/ProposedMigration#Reproducing_tests_in_the_cloud) is quite similar to the other invocations. Just two things change compared to "local" autopkgtest-runner invocations.
+with `nova` and adding the `setup-testbed` setup command.
```
If you intended to keep this in a separate line, you should leave another line. Otherwise, `with` -> `With`, at least.

cpaelzer replied:
Thanks, fixed
Now that we have the bullet list afterwards this line was no more necessary anyway

utkarsh2102 commented on PackageTests.md:
```diff
@@ -78,6 +78,54 @@ The command only differs after the `--` part. For example:
+
+ * `--setup-commands setup-testbed` will have autopkgtest execute `/usr/share/autopkgtest/setup-commands/setup-testbed` on the target which converts any system into a system that is ready for autopkgtest to log in
+ * `-- ssh -s nova` achieves two things
+  * First of all it selects the ssh virtualization driver `autopkgtest-virt-ssh` to reach out to a remote system
+  * Furthermore it selects the setup script `nova` from `/usr/share/autopkgtest/ssh-setup/nova` which happens to know how to deal with openstack
```
MD formatting issues against, unfortunately; see: https://github.com/cpaelzer/ubuntu-maintainers-handbook/blob/extend-autopkgtest/PackageTests.md

Instead of the 2nd point having 2 more points, it looks like there are 4 points in all, which doesn't seem right, right? :)

utkarsh2102 commented on PackageTests.md:
```diff
@@ -78,6 +78,54 @@ The command only differs after the `--` part. For example:
+
+Quite often a test fails by running against new packages in the proposed pocket. Then it often will be helpful to check if the test needs other packages from proposed to resolve the issue. That can easily be done via the option `--apt-pocket`.
+
+Commonly a test will run against all packages in release plus the new candidate from proposed, that would look like:
+
+    --apt-pocket=proposed=src:yourpkg
+
+To run against all packages that are in proposed, you'd simply not refer to a package
+
+    --apt-pocket=proposed
+
+And if instead you'd need a given set of packages, but not everything else from proposed you can use a comma separated list
+
+    --apt-pocket=proposed=src:srcpkg1,srcpkg2
```
Maybe worth adding an example like done for the above paragraph? But fine, either way.

cpaelzer replied:
Done and along that also added how to size a VM

utkarsh2102 reviewed (changes requested):
Great work, @cpaelzer. This looks great! 🎉 
I just have few minor comments below.

cpaelzer commented:
Addressed the feedback of @bryceharrington and @utkarsh2102 now - anything else?

cpaelzer commented:
We have had two reviews (thanks!) and I've addressed all feedback.
I need to close out some of my concurrent tasks and would now merge this as-is.
If there are findings later it is easy enough to further improve the documentation.

bryceharrington commented:
On Wed, Mar 24, 2021 at 11:53:44PM -0700, Christian Ehrhardt wrote:
> The fixups for @bryce made me made me also improve a few other things here and there.
> Thanks for the feedback, I've pushed my cleanups to this branch.
> Is there anything else that would need to change or can we merge it this way now?

Just a couple markdown formatting things:

Looking at this page, the code snippets need to be set as pre-formatted
since the command lines starting at the `source` line are wrapping:

   https://github.com/cpaelzer/ubuntu-maintainers-handbook/blob/extend-autopkgtest/PackageTests.md

The two bullets under `ssh -s nova` need indented one more level.

utkarsh2102 commented:
I mentioned that in my review, maybe Christian missed it. But that's okay, no problem. I'll fix that, amongst other things.
